### PR TITLE
DOMA-1474 [FRONTEND] resident and organization ticket types

### PR DESCRIPTION
**Pull request by nomerdvadcatpyat**
- added `type` field to `TicketComment` schema (`organization` and `resident` options)
- updated Comments markup
- wrote access rights to `TicketComment` for resident cases
- wrote tests

### CODE REVIEW

File: apps/condo/domains/ticket/access/TicketComment.js [hunk 1]
```diff
@@ -7,14 +7,44 @@ const { throwAuthenticationError } = require('@condo/domains/common/utils/apollo
...
     if (user.isSupport || user.isAdmin) return {}
 
+    if (user.type === RESIDENT) {
+        const residents = await find('Resident', { user: { id: user.id }, deletedAt: null })
+
+        if (isEmpty(residents)) return false
+
+        const organizationsIds = compact(residents.map(resident => get(resident, 'organization')))
+        const residentAddressOrStatement = residents.map(resident =>
+            ({ AND: [{ canReadByResident: true, contact: { phone: user.phone } }, { property: { id: resident.property } }, { unitName: resident.unitName }] }))
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketComment.js:31]: what about unitType ?

File: apps/condo/domains/ticket/access/TicketComment.js [hunk 2]
```diff
@@ -32,21 +62,63 @@ async function canManageTicketComments ({ authentication: { item: user }, origin
...
+        const residents = await find('Resident', {
+            user: { id: user.id },
+            property: { id: propertyId, deletedAt: null },
+            unitName,
+            deletedAt: null,
+        })
+
+        return residents.length > 0
+    }
+    else if (user.type === STAFF) {
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketComment.js:105]:
```suggestion
    } else if (user.type === STAFF) {
```

File: apps/condo/domains/ticket/access/TicketComment.js [hunk 3]
```diff
@@ -32,21 +62,63 @@ async function canManageTicketComments ({ authentication: { item: user }, origin
...
+            return false
+        }
+
+        const propertyId = get(ticket, 'property', null)
+        const unitName = get(ticket, 'unitName', null)
+
+        const residents = await find('Resident', {
+            user: { id: user.id },
+            property: { id: propertyId, deletedAt: null },
+            unitName,
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketComment.js:99]: unitType

File: apps/condo/domains/ticket/schema/TicketComment.js [hunk 4]
```diff
@@ -47,7 +56,17 @@ const TicketComment = new GQLListSchema('TicketComment', {
...
-    plugins: [uuided(), versioned(), tracked(), softDeleted(), historical()],
+    plugins: [
+        uuided(), versioned(),
+        tracked({
+            relationshipAccess: {
+                read: access.canReadUserField,
+                update: false,
+                create: false,
+            },
+        }),
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketComment.js:67]: Why do you need to customize it?

File: apps/condo/domains/ticket/schema/TicketComment.js [hunk 5]
```diff
@@ -2,11 +2,12 @@
  * Generated by `createschema ticket.TicketComment 'ticket:Relationship:Ticket:CASCADE; user:Relationship:User:CASCADE; content:Text;'`
  */
 
-const { Text, Relationship } = require('@keystonejs/fields')
+const { Text, Relationship, Select } = require('@keystonejs/fields')
 const { GQLListSchema } = require('@core/keystone/schema')
 const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
 const { SENDER_FIELD, DV_FIELD } = require('@condo/domains/common/schema/fields')
 const access = require('@condo/domains/ticket/access/TicketComment')
+const { COMMENT_TYPE } = require('@condo/domains/ticket/constants')
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketComment.js:10]: looks inconsistent. Look at User.type

File: apps/condo/domains/ticket/schema/TicketComment.test.js [hunk 6]
```diff
@@ -15,370 +15,940 @@ const { makeClientWithProperty } = require('@condo/domains/property/utils/testSc
...
             })
-            const [ticket] = await createTestTicket(admin, organizationFrom, propertyFrom)
-            const [comment] = await createTestTicketComment(admin, ticket, clientFrom.user)
-            const payload = {
-                content: faker.random.alphaNumeric(10),
-            }
 
-            await expectToThrowAccessDeniedErrorToObj(async () => {
-                await updateTestTicketComment(clientTo, comment.id, payload)
+            it('cannot read comments in not his resident ticket', async () => {
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketComment.test.js:656]: We can also check the read after Resident.deletedAt. And read after Property.deletedAt and Ticket.deletedAt

File: apps/condo/domains/ticket/utils/testSchema/index.js [hunk 7]
```diff
@@ -8,7 +8,7 @@ const faker = require('faker')
 const { get } = require('lodash')
 const { makeClientWithProperty } = require('@condo/domains/property/utils/testSchema')
 const { makeLoggedInAdminClient } = require('@core/keystone/test.utils')
-const { TICKET_STATUS_TYPES } = require('../../constants')
+const { TICKET_STATUS_TYPES, COMMENT_TYPE} = require('../../constants')
```
Comment (pahaz) [inline apps/condo/domains/ticket/utils/testSchema/index.js:11]:
```suggestion
const { TICKET_STATUS_TYPES, COMMENT_TYPE } = require('../../constants')
```

File: apps/condo/schema.ts [hunk 8]
```diff
@@ -27654,7 +27687,7 @@ export enum SendMessageType {
   ShareTicket = 'SHARE_TICKET',
   DeveloperImportantNoteType = 'DEVELOPER_IMPORTANT_NOTE_TYPE',
   CustomerImportantNoteType = 'CUSTOMER_IMPORTANT_NOTE_TYPE',
-  MessageForwardedToSupport = 'MESSAGE_FORWARDED_TO_SUPPORT_TYPE',
+  MessageForwardedToSupport = 'MESSAGE_FORWARDED_TO_SUPPORT',
```
Comment (pahaz) [inline apps/condo/schema.ts:27690]: conflict! need to rebase

File: packages/@core.keystone/plugins/tracked.js [hunk 9]
```diff
@@ -1,7 +1,7 @@
 const { composeHook } = require('./utils')
 const { plugin } = require('./utils/typing')
 
-const tracked = ({ createdAtField = 'createdAt', createdByField = 'createdBy', updatedAtField = 'updatedAt', updatedByField = 'updatedBy' } = {}) => plugin(({ fields = {}, hooks = {}, ...rest }) => {
+const tracked = ({ createdAtField = 'createdAt', createdByField = 'createdBy', relationshipAccess, updatedAtField = 'updatedAt', updatedByField = 'updatedBy' } = {}) => plugin(({ fields = {}, hooks = {}, ...rest }) => {
```
Comment (pahaz) [inline packages/@core.keystone/plugins/tracked.js:4]: I don't think that we need to customise it

File: apps/condo/schema.ts (hunk 8 above)
Comment (nomerdvadcatpyat) [reply apps/condo/schema.ts:27690]: rebased, made types again
it's still here

File: packages/@core.keystone/plugins/tracked.js (hunk 9 above)
Comment (nomerdvadcatpyat) [reply packages/@core.keystone/plugins/tracked.js:4]: we need to hide from resident ticket author
изображение

File: apps/condo/domains/ticket/schema/TicketComment.js (hunk 4 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/schema/TicketComment.js:67]: we need to hide ticket author from resident
изображение

**Review by dkoviazin — APPROVED** [review]

**Review by sitozzz — APPROVED** [review]

File: apps/condo/domains/ticket/schema/TicketComment.js (hunk 4 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/ticket/schema/TicketComment.js]: removed it

File: packages/@core.keystone/plugins/tracked.js (hunk 9 above)
Comment (nomerdvadcatpyat) [reply packages/@core.keystone/plugins/tracked.js]: removed it

File: apps/condo/domains/ticket/access/TicketComment.js [hunk 10]
```diff
@@ -32,21 +60,64 @@ async function canManageTicketComments ({ authentication: { item: user }, origin
...
+            unitType,
+            deletedAt: null,
+        })
+
+        return residents.length > 0
+    } else if (user.type === STAFF) {
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketComment.js:104]: !! Will not work with service user type

**Review by pahaz — APPROVED** [review]

**Pull request closed without merging**